test/functorch/test_aotdispatch.py

```diff
@@ -6022,6 +6022,162 @@ def fn(x):
+
+        ref_inps = dense_inps()
+        ref_outs = m(*ref_inps)
+        ref_outs[0].sum().backward()
+
+        inps = dense_inps()
+        outs = torch.compile(m, backend="inductor", fullgraph=True)(*inps)
+        s = outs[0].sum()
+        # Using torch.profiler to verify that no contiguous() calls happened
+        with torch.profiler.profile(
+            activities=[torch.profiler.ProfilerActivity.CPU],
+        ) as prof:
+            s.backward()
+
+        contiguous_events = [
+            event for event in prof.events() if event.name == "aten::contiguous"
+        ]
+        self.assertTrue(len(contiguous_events) == 0)
+
+    def test_channels_last_grads_no_force_contiguous_subclass(self):
+        class M(torch.nn.Module):
+            def __init__(self) -> None:
+                super().__init__()
+                self.conv = torch.nn.Conv2d(3, 3, 3)
+
+            def forward(self, x, y):
+                r = self.conv(x)
+                return r, y + 1
+
+        m = M()
+        m.to(memory_format=torch.channels_last)
+        m.train()
+
+        def inps_fn():
+            return (
+                TwoTensor(
+                    torch.randn(2, 3, 5, 5, requires_grad=True).to(
+                        memory_format=torch.channels_last
+                    ),
+                    torch.randn(2, 3, 5, 5, requires_grad=True).to(
+                        memory_format=torch.channels_last
+                    ),
+                ),
+                torch.randn(3, 2, requires_grad=True).clone(),
+            )
+
+        ref_inps = inps_fn()
+        ref_outs = m(*ref_inps)
+        ref_outs[0].sum().backward()
+
+        mc = M()
+        mc.to(memory_format=torch.channels_last)
+        mc.train()
+        inps = inps_fn()
+        outs = torch.compile(mc, backend="aot_eager", fullgraph=True)(*inps)
+        s = outs[0].sum()
+        with torch.profiler.profile(
+            activities=[torch.profiler.ProfilerActivity.CPU],
+        ) as prof:
+            s.backward()
+
+        contiguous_events = [
+            event for event in prof.events() if event.name == "aten::contiguous"
+        ]
+        self.assertTrue(len(contiguous_events) == 0)
+
+    def test_channels_last_grads_no_force_contiguous_nested_subclass(self):
+        class M(torch.nn.Module):
+            def __init__(self) -> None:
+                super().__init__()
+                self.conv = torch.nn.Conv2d(3, 3, 3)
+
+            def forward(self, x, y):
+                r = self.conv(x)
+                return r, y + 1
+
+        m = M()
+        m.to(memory_format=torch.channels_last)
+        m.train()
+
+        def inps_fn():
+            return (
+                TwoTensor(
+                    TwoTensor(
+                        torch.randn(2, 3, 5, 5, requires_grad=True).to(
+                            memory_format=torch.channels_last
+                        ),
+                        torch.randn(2, 3, 5, 5, requires_grad=True).to(
```

Comment:
even better, let's make the inner two tensors have a mix of channels-first-contiguous and channels-last-contiguous, and ensure that we **still** don't have any runtime contiguous calls (so we "guessed all of our memory formats" properly)